Upgrade to `pulumi/pulumi` 3.120.0

lunaris opened the pull request:
This commit bumps the bridge to use the latest available version of Pulumi core. The most notable change here is that provider method signatures now accept a `context.Context`, meaning that we can deprecate the bridge's forked `ProviderWithContext` interface and methods.

lunaris commented on pkg/tfgen/pluginHost.go:
```diff
@@ -122,7 +138,7 @@ func (host *inmemoryProviderHost) Provider(pkg tokens.Package, version *semver.V
 	version *semver.Version) (*workspace.PluginInfo, error) {
 	if name == host.provider.name.String() {
-		info, err := host.provider.GetPluginInfo()
+		info, err := host.provider.GetPluginInfo(context.TODO())
```
`context.Background()` is also an option here, potentially -- thoughts? I'm not sure if it's possible to change the signature of this method/what the implications of that are, but that's also an option.

lunaris commented on pkg/tfbridge/provider.go:
```diff
@@ -669,7 +669,25 @@ func DiffConfig(
+		allowUnknowns bool, ignoreChanges []string,
+	) (plugin.DiffResult, error) {
+		return differ.DiffConfig(
+			context.TODO(),
```
I've done this to avoid changing the signature of this function to accept a context, because I assume that will break other things, but happy to do that, or change this to e.g. `context.Background()`.

VenelinMartinov replied:
I'd guess this is ok to change the signature since it's only used in pf and that already has a context.

I don't imagine this being used externally.

lunaris commented:
Apologies, I completely missed `pf`. This is interesting because currently we have `ProviderWithContext`, but now that all the methods have been standardised to take a `Context`, we might be able to do away with this entire concept, maybe?

iwahbe commented:
> Apologies, I completely missed `pf`. This is interesting because currently we have `ProviderWithContext`, but now that all the methods have been standardised to take a `Context`, we might be able to do away with this entire concept, maybe?

Making an internal thread public:

> ProviderWithContext can and should be removed with the new signatures. I originally redid plugin.Provider because I was tired of plumbing changes through ProviderWithContext. The bridge actually has a fork of some of plugin ([here](https://github.com/pulumi/pulumi-terraform-bridge/blob/e1c41e3501d86de626ebf2e0af07a1ba6df33424/pf/internal/plugin/package.go#L15-L25)), which I’m hoping we can now remove.
>
>Any PR that you open which gets the bridge compiling against latest pu/pu would be welcome! That said, the more of the plugin folder you can remove from the bridge, the better.

lunaris commented:
I've made some progress on this, but probably still need a few pointers. I've removed `pf/internal/plugin` and made `provider` an instance of `plugin.Provider` by tweaking the method signatures appropriately. This all appears to type check fine. However, the tests fail. I suspect this is down to two things:

1. I've made a mistake when porting the old signatures to the new (e.g. passed old inputs instead of old outputs or similar due to differences in naming). I've spotted and fixed one such bug already (thanks tests!) but might have missed others.
2. The internal code used a bespoke `configEncode` to handle unmarshalling in `CheckConfig`/`DiffConfig`. I don't know if there is a natural way to port this behaviour under `plugin.Provider` -- if there is, can someone advise? If not, I guess we will be unable to use `plugin.NewProviderServer` as-is and have to wrap it still?

Of course it's possible there are other problems/things I've missed -- all input welcome!

iwahbe reviewed (commented):
It looks like tests are failing here:

```
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -758,3 +758,3 @@
        	            	     ),
        	            	-    opts=pulumi.ResourceOptions(depends_on=[alpha]))
        	            	+    opts = pulumi.ResourceOptions(depends_on=[alpha]))
        	            	 ```
        	            	@@ -1119,3 +1119,3 @@
        	            	 ),
        	            	-opts=pulumi.ResourceOptions(depends_on=[
        	            	+opts = pulumi.ResourceOptions(depends_on=[
        	            	         lambda_logs,
        	Test:       	TestConvertExamples/aws_lambda_function
--- FAIL: TestConvertExamples (13.45s)
    --- PASS: TestConvertExamples/wavefront_dashboard_json/setup (1.78s)
    --- PASS: TestConvertExamples/wavefront_dashboard_json (0.11s)
    --- PASS: TestConvertExamples/golang_wavefront_dashboard_json/setup (0.16s)
    --- PASS: TestConvertExamples/golang_wavefront_dashboard_json (0.04s)
    --- PASS: TestConvertExamples/equinix_fabric_connection/setup (1.74s)
    --- PASS: TestConvertExamples/equinix_fabric_connection (0.00s)
    --- PASS: TestConvertExamples/aws_lambda_function/setup (6.91s)
    --- FAIL: TestConvertExamples/aws_lambda_function (2.71s)
```

That's due to a change in convert formatting and is not a problem in plugin mapping. It's just a changed introduced in 3.120.0 (separate from the `plugin.Provider` interface).

---

Let's get CI green (fix test & fix lints), then come back to the encoding. If custom config encoders are not needed to pass CI, then we'll need to take a deeper look. It could be either:

- Custom config encoders are important, and not under test. To merge this PR, we need to add tests (in a separate PR) that verify custom config encoder behavior and then rebase this PR (fixing the encoder so the new tests pass).
- Custom config encoders are not important. We can remove any mention of them from this PR and move on.

VenelinMartinov commented:
I think @iwahbe opened https://github.com/pulumi/pulumi-terraform-bridge/issues/2148 as a follow-up
